source/Scrapers/TeamCityCompatibleAgentsScraper.cs

```diff
@@ -0,0 +1,70 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Configuration;
+using Prometheus.Client;
+using Serilog;
+using TeamCitySharp;
+
+namespace TeamCityBuildStatsScraper.Scrapers
+{
+    class TeamCityCompatibleAgentsScraper : BackgroundService
+    {
+        readonly IMetricFactory metricFactory;
+        readonly IConfiguration configuration;
+        readonly HashSet<(string buildTypeId, string buildId)> seenBuildsNoAgents = new();
+
+        public TeamCityCompatibleAgentsScraper(IMetricFactory metricFactory, IConfiguration configuration, ILogger logger)
+            : base(logger.ForContext("Scraper", nameof(TeamCityCompatibleAgentsScraper)))
+        {
+            this.metricFactory = metricFactory;
+            this.configuration = configuration;
+        }
+
+        protected override TimeSpan DelayBetweenScrapes => TimeSpan.FromSeconds(60);
+
+        protected override async Task Scrape(CancellationToken stoppingToken)
+        {
+            await Task.CompletedTask;
+
+            var teamCityToken = configuration.GetValue<string>("TEAMCITY_TOKEN");
+            var teamCityUrl = configuration.GetValue<string>("BUILD_SERVER_URL");
+            var useSSL = configuration.GetValue<bool>("USE_SSL");
+            var teamCityClient = new TeamCityClient(teamCityUrl, useSSL);
+
+            teamCityClient.ConnectWithAccessToken(teamCityToken);
+
+            var queuedBuilds = teamCityClient.BuildQueue
+                .GetFields("count,build(id,waitReason,buildTypeId,queuedDate,statistics(property,value,name),compatibleAgents(count,agent(id)))")
+                .All()
+                // exclude builds with no wait reason - these are the ones that are 'starting shortly'
+                .Where(qb => qb.WaitReason != null)
+                .ToArray();
+
+            // Track builds with no compatible agents
+            var buildsNoCompatibleAgents = queuedBuilds
+                .Where(qb => qb.WaitReason == "There are no idle compatible agents which can run this build")
+                .Where(qb => qb.CompatibleAgents?.Agent == null || qb.CompatibleAgents.Agent.Count == 0)
+                .ToArray();
+
+            var noAgentsGauge = metricFactory.CreateGauge("queued_builds_no_compatible_agents", "Queued builds waiting with no compatible agents available", "buildTypeId", "buildId", "queuedDateTime");
```

Comment:
We spoke on Zoom about how having `buildId` in here impacts cardinality; with the expected magnitude of results I don't think it's a huge problem in this case.